exec/local.go

```diff
@@ -64,6 +64,41 @@ func (l *localExecutor) Run(task *Task) {
 		return
 	}
 	defer l.limiter.Release(n)
+	in, err := l.depReaders(ctx, task)
+	if err != nil {
+		if errors.Match(fatalErr, err) {
+			task.Error(err)
+		} else {
+			task.Set(TaskLost)
+		}
+		return
+	}
+	task.Set(TaskRunning)
+
+	// Start execution, then place output in a task buffer. We also plumb a
+	// metrics scope in here so we can store and aggregate metrics.
+	task.Scope.Reset(nil)
+	out := task.Do(in)
+	buf, err := bufferOutput(metrics.ScopedContext(ctx, &task.Scope), task, out)
+	task.Lock()
+	if err == nil {
+		l.mu.Lock()
+		l.buffers[task] = buf
+		l.mu.Unlock()
+		task.state = TaskOk
+	} else {
+		if errors.Match(fatalErr, err) {
+			task.state = TaskErr
+		} else {
+			task.state = TaskLost
```

Comment:
Prior to this PR, there was no way for tasks to become lost with the local evaluator. Now, discarding can lead to lost tasks. This means that reads from dependencies can now fail when trying to read from lost tasks. This change makes it so that read errors are not considered fatal. The restructuring was to try to make clearer the logic for deciding error vs. lost tasks by making the treatment of reads done for input combining similar to the ones done by normal task execution, i.e. the result of `bufferOutput`. Maybe it only made it confusing?

Reply:
I didn't realize there was no lost path for the local evaluator before this. Now, I don't think it's confusing.